### PR TITLE
[nrf noup] Wifi directed scan with single scan

### PR DESCRIPTION
**Pull request by LipinskiPNordicSemi**
This commit implements directed scanning and single ssid scanning while commissioning.

### CODE REVIEW

File: src/platform/nrfconnect/wifi/WiFiManager.cpp [hunk 1]
```diff
@@ -188,14 +188,26 @@ CHIP_ERROR WiFiManager::Scan(const ByteSpan & ssid, ScanResultCallback resultCal
     mWiFiState          = WIFI_STATE_SCANNING;
     mSsidFound          = false;
 
-    if (0 != net_mgmt(NET_REQUEST_WIFI_SCAN, mNetIf, NULL, 0))
+    if (!ssid.empty())
     {
-        ChipLogError(DeviceLayer, "Scan request failed");
-        return CHIP_ERROR_INTERNAL;
+        memcpy(mScanSsidBuffer, ssid.data(), ssid.size());
+        mScanParams.ssids[0] = reinterpret_cast<const char *>(&mScanSsidBuffer);
```
Comment (Damian-Nordic) [inline src/platform/nrfconnect/wifi/WiFiManager.cpp:194]: I think SSID and scan params could be allocated on stack. Also, make sure you allocate enough buffer for null-terminator and validate the input string size:

```suggestion
        char ssidBuffer[DeviceLayer::Internal::kMaxWiFiSSIDLength + 1] = {};
        wifi_scan_params scanParams = {};
        
        VerifyOrReturnError(ssid.size() < sizeof(ssidBuffer), CHIP_ERROR_INVALID_ARGUMENT);
        memcpy(ssidBuffer, ssid.data(), ssid.size());
        scanParams.ssid[0] = ssidBuffer;
```

**Review by Damian-Nordic — APPROVED** [review]
I think this is a better approach than filtering results in callback.

File: src/platform/nrfconnect/wifi/WiFiManager.cpp (hunk 1 above)
Comment (LipinskiPNordicSemi) [reply src/platform/nrfconnect/wifi/WiFiManager.cpp:194]: The problem with those variables on the stac is that I think the function is finished before scanning (scanning might be done on the zephyr thead).
If I try directed scanning with your suggestions, the result is like without SSID given.

Comment (LipinskiPNordicSemi) [reply src/platform/nrfconnect/wifi/WiFiManager.cpp:194]: After implementing suggested changes, I found some problems with directed scanning - one in few scan couldn't find network by ssid

Comment (Damian-Nordic) [reply src/platform/nrfconnect/wifi/WiFiManager.cpp:194]: ok, nevermind then but please validate the SSID length in your solution and add `+1` to the buffer size :)

**Review by LuDuda — APPROVED** [review]
Looks great!

File: src/platform/nrfconnect/wifi/WiFiManager.cpp (hunk 1 above)
Comment (markaj-nordic) [reply src/platform/nrfconnect/wifi/WiFiManager.cpp]: Looks like the Wi-Fi driver uses only the shallow copy of `wifi_scan_params`, so we cannot rely on stack allocated objects here.

File: src/platform/nrfconnect/wifi/WiFiManager.cpp [hunk 2]
```diff
@@ -188,7 +188,18 @@ CHIP_ERROR WiFiManager::Scan(const ByteSpan & ssid, ScanResultCallback resultCal
...
 
-    if (0 != net_mgmt(NET_REQUEST_WIFI_SCAN, mNetIf, NULL, 0))
+    wifi_scan_params * scanParamsFinal{ nullptr };
+
+    if (!ssid.empty())
+    {
+        /* We must assume that the ssid is handled as a NULL-terminated string.
+           Note that the mScanSsidBuffer is initialized with zeros. */
+        VerifyOrReturnError(ssid.size() < sizeof(mScanParams), CHIP_ERROR_INVALID_ARGUMENT);
+        memcpy(mScanSsidBuffer, ssid.data(), ssid.size());
```
Comment (Damian-Nordic) [inline src/platform/nrfconnect/wifi/WiFiManager.cpp:198]: Can you add null-terminator here? Thinking about the case in which we first scan for "my-network-1" and then some shorter "my-network".

File: src/platform/nrfconnect/wifi/WiFiManager.cpp [hunk 3]
```diff
@@ -188,7 +188,18 @@ CHIP_ERROR WiFiManager::Scan(const ByteSpan & ssid, ScanResultCallback resultCal
...
+        mScanParams.ssids[0] = mScanSsidBuffer;
+        scanParamsFinal      = &mScanParams;
+    }
+    if (0 != net_mgmt(NET_REQUEST_WIFI_SCAN, mNetIf, scanParamsFinal, sizeof(*scanParamsFinal)))
```
Comment (Damian-Nordic) [inline src/platform/nrfconnect/wifi/WiFiManager.cpp:202]: Is it OK to pass non-zero size even if params are null?

**Review by Damian-Nordic — APPROVED** [review]

File: src/platform/nrfconnect/wifi/WiFiManager.cpp (hunk 1 above)
Comment (Damian-Nordic) [reply src/platform/nrfconnect/wifi/WiFiManager.cpp]: Thanks for checking!

File: src/platform/nrfconnect/wifi/WiFiManager.cpp [hunk 4]
```diff
@@ -188,7 +188,18 @@ CHIP_ERROR WiFiManager::Scan(const ByteSpan & ssid, ScanResultCallback resultCal
...
-    if (0 != net_mgmt(NET_REQUEST_WIFI_SCAN, mNetIf, NULL, 0))
+    wifi_scan_params * scanParamsFinal{ nullptr };
+
+    if (!ssid.empty())
+    {
+        /* We must assume that the ssid is handled as a NULL-terminated string.
+           Note that the mScanSsidBuffer is initialized with zeros. */
+        VerifyOrReturnError(ssid.size() < sizeof(mScanParams), CHIP_ERROR_INVALID_ARGUMENT);
+        memcpy(mScanSsidBuffer, ssid.data(), ssid.size());
+        mScanParams.ssids[0] = mScanSsidBuffer;
```
Comment (krish2718) [inline src/platform/nrfconnect/wifi/WiFiManager.cpp:199]: as there is no length of SSIDs, the driver relies on `ssids[1] = NULL` to know the length.

File: src/platform/nrfconnect/wifi/WiFiManager.cpp (hunk 3 above)
Comment (markaj-nordic) [reply src/platform/nrfconnect/wifi/WiFiManager.cpp:202]: I think so. But, ok, let's be more cautious.

File: src/platform/nrfconnect/wifi/WiFiManager.cpp (hunk 4 above)
Comment (markaj-nordic) [reply src/platform/nrfconnect/wifi/WiFiManager.cpp:199]: I see, shall we set the max_bss_cnt to 1 in such a case?

File: src/platform/nrfconnect/wifi/WiFiManager.cpp (hunk 2 above)
Comment (markaj-nordic) [reply src/platform/nrfconnect/wifi/WiFiManager.cpp:198]: Valid point, thanks.

File: src/platform/nrfconnect/wifi/WiFiManager.cpp (hunk 4 above)
Comment (krish2718) [reply src/platform/nrfconnect/wifi/WiFiManager.cpp:199]: no, `max_bss_cnt` is for limiting scan results, it's not used for limiting SSIDs

**Comment by markaj-nordic** [comment]
@krish2718, @Damian-Nordic - please check one last time.

File: src/platform/nrfconnect/wifi/WiFiManager.cpp [hunk 5]
```diff
@@ -188,7 +188,22 @@ CHIP_ERROR WiFiManager::Scan(const ByteSpan & ssid, ScanResultCallback resultCal
     mWiFiState          = WIFI_STATE_SCANNING;
     mSsidFound          = false;
 
-    if (0 != net_mgmt(NET_REQUEST_WIFI_SCAN, mNetIf, NULL, 0))
+    wifi_scan_params * scanParamsFinal{ nullptr };
+    size_t scanParamsLen{ 0 };
```
Comment (LuDuda) [inline src/platform/nrfconnect/wifi/WiFiManager.cpp:192]:
```suggestion
    size_t scanParamsSize{ 0 };
```

File: src/platform/nrfconnect/wifi/WiFiManager.cpp [hunk 6]
```diff
@@ -188,7 +188,22 @@ CHIP_ERROR WiFiManager::Scan(const ByteSpan & ssid, ScanResultCallback resultCal
     mWiFiState          = WIFI_STATE_SCANNING;
     mSsidFound          = false;
 
-    if (0 != net_mgmt(NET_REQUEST_WIFI_SCAN, mNetIf, NULL, 0))
+    wifi_scan_params * scanParamsFinal{ nullptr };
```
Comment (LuDuda) [inline src/platform/nrfconnect/wifi/WiFiManager.cpp:191]:
```suggestion
    wifi_scan_params * scanParams{ nullptr };
```

**Review by LuDuda — APPROVED** [review]

**Review by Damian-Nordic — APPROVED** [review]
Looks great now

**Pull request merged**